server/keyspace/keyspace.go

```diff
@@ -0,0 +1,240 @@
+var (
+	errKeyspaceArchived   = errors.New("Keyspace already archived")
+	ErrKeyspaceNotFound   = errors.New("Keyspace does not exist")
+	ErrKeyspaceNameExists = errors.New("Keyspace name already in use")
+	errIllegalID          = errors.New("Cannot create keyspace with that ID")
+	errIllegalName        = errors.New("Cannot create keyspace with that name")
+)
+
+// Manager manages keyspace related data.
+// It validates requests and provides concurrency control.
+type Manager struct {
+	// idLock guards keyspace name to id lookup entries.
+	idLock syncutil.Mutex
+	// metaLock guards keyspace meta.
+	metaLock syncutil.Mutex
+	// idAllocator allocates keyspace id.
+	idAllocator id.Allocator
+	// store is the storage for keyspace related information.
+	store endpoint.KeyspaceStorage
+}
+
+// CreateKeyspaceRequest represents necessary arguments to create a keyspace.
+type CreateKeyspaceRequest struct {
+	Name          string
+	InitialConfig map[string]string
+	Now           time.Time
+}
+
+// UpdateKeyspaceRequest represents necessary arguments to update keyspace.
+type UpdateKeyspaceRequest struct {
+	Name string
+	// Whether to update the state of keyspace,
+	// if set to false, NewState and Now will be ignored.
+	UpdateState bool
+	NewState    keyspacepb.KeyspaceState
+	Now         time.Time
+	// New KVs to put in keyspace config.
+	ToPut map[string]string
+	// KVs to remove from keyspace config.
+	// Removal will be performed after put.
+	ToDelete []string
+}
+
+// NewKeyspaceManager creates a Manager of keyspace related data.
+func NewKeyspaceManager(store endpoint.KeyspaceStorage, idAllocator id.Allocator) *Manager {
+	return &Manager{
+		store:       store,
+		idAllocator: idAllocator,
+	}
+}
+
+// allocID allocate a new keyspace id.
+func (manager *Manager) allocID() (uint32, error) {
+	id64, err := manager.idAllocator.Alloc()
+	if err != nil {
+		return 0, err
+	}
+	id32 := uint32(id64)
+	// skip id 0
+	if id32 < spaceIDMin {
+		return manager.allocID()
+	}
+	// if allocated id too big, return error
+	if id32 > spaceIDMax {
+		return 0, errIllegalID
+	}
+	return id32, nil
+}
+
+// createNameToID create a keyspace name to ID lookup entry.
+// It returns error if saving keyspace name meet error or if name has already been taken.
+func (manager *Manager) createNameToID(spaceID uint32, name string) error {
+	manager.idLock.Lock()
+	defer manager.idLock.Unlock()
+
+	nameExists, _, err := manager.store.LoadKeyspaceID(name)
+	if err != nil {
+		return err
+	}
+	if nameExists {
+		return ErrKeyspaceNameExists
+	}
+	return manager.store.SaveKeyspaceID(spaceID, name)
+}
+
+// CreateKeyspace create a keyspace meta with initial config and save it to storage.
+func (manager *Manager) CreateKeyspace(request *CreateKeyspaceRequest) (*keyspacepb.KeyspaceMeta, error) {
+	// allocate new id
+	newID, err := manager.allocID()
+	if err != nil {
+		return nil, err
+	}
+	// bind name to that id
+	if err = manager.createNameToID(newID, request.Name); err != nil {
```

Comment:
Name to id and meta should be updated in a transaction. Other wise user could see name to id mapping without metadata associated. 

For the case if we really can't use transaction, we need to reorder updates and make sure user can't see it before it's fully initialized and collect garbages.